### PR TITLE
Adagrams Submission - Danielle & Alice

### PR DESCRIPTION
**Pull request by danimetz**
# Adagrams
Congratulations! You're submitting your assignment.

## Comprehension Questions

Feature | Feedback
--- | ---
What are the components that make up a method? |  Method signature (optional parameters),  method body - block of code,  explicit or implicit return|
What are the advantages of using git when collaboratively working on one code base? | You can simultaneously work on the code, by pushing and pulling from github, and it will keep track of all the different versions and commit messages.|
What kind of relationship did you and your pair have with the unit tests? | It allowed us to understand our expected outputs for our methods. If there were errors, we could then isolate where our program was failing.|
Does your code use any methods from the `Enumerable` mixin? If so, where and why was it helpful? | .all? was used to make sure all the letters were in the hand. .max_by was used to determine what the max score was. .select was used to isolate values that had the max score. Min_by was used to determine the minimum length of the words that had max scores. Then we used .select again to isolate words & scores that had the max score but were also the shortest or 10 letters long. We also used include? to determine if the words were in the english dictionary. |
What was one method you and your pair used to debug code? |We ran adagram.rb and used puts statements to see what was being returned. We also used .class to determine the structure of the data being returned so we could access the data later in our code.|
What are two discussion points that you and your pair discussed when giving/receiving feedback from each other that you would be willing to share? |There was a good balance of give and take. We were both flexible when it came to sharing ideas and strategies and were willing to try different ways for solving each of the waves.|

### CODE REVIEW

File: lib/adagrams.rb [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+      value -= 1
+    end
+  end
+
+  #Select letters and replace with 0 once they are picked
+  your_hand = []
+  count = 0
+  while count < 10 do
+    index = rand(0...letter_probability.length)
+    if letter_probability[index] != 0
```
Comment (dHelmgren) [inline lib/adagrams.rb:52]: So, this works, but you end up with an array that becomes more and more likely to "bounce" as you pick letters from the pile. Instead of nulling the value at the index, why not remove just that element from the array? That way, you don't waste loops through on no-ops (no operation).

File: lib/adagrams.rb [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+#Wave 2
+#Checks that the letters for the word you picked are in your hand
+def uses_available_letters?(input, letters_in_hand)
+  input = input.upcase.split("")
+  value = []
+  if input.length > letters_in_hand.length
+    return false
+  else
+    input.each do |letter|
+      input.count(letter) > letters_in_hand.count(letter) ? value << false : value << true
```
Comment (dHelmgren) [inline lib/adagrams.rb:70]: You are doing more work than you need to! As soon as one of these checks comes up false, you can return false, right? The only case that requires you to complete the loop is if every letter matches!

File: lib/adagrams.rb [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+    end
+  end
+  value.all? true ? true : false
```
Comment (dHelmgren) [inline lib/adagrams.rb:73]: `value.all?` is hiding a second loop from you!

**Comment by dHelmgren** [comment]
# Adagrams
## What We're Looking For

Feature | Feedback
--- | ---
**General** |
Answered comprehension questions | Good, comprehensive answers
Both teammates contributed to the codebase | +
Small commits with meaningful commit messages | Commits were lean and clean! 
**Code Requirements** |
`draw_letters` method | It works, but see my comment!
Uses appropriate data structure to store the letter distribution | Conversion from a hash to an array was easy to miss, but I can understand why you opted for that!
All tests for `draw_letters` pass | Yes indeed!
`uses_available_letters?` method | Again, good, but check the comments! 
All tests for `uses_available_letters?` pass | Green across the board!
`score_word` method | * Big Thumbs Up *
Uses appropriate data structure to store the letter scores | So, it's not in a data structure, but your case statement is good!
All tests for `score_word` pass | P-P-P-PASS
`highest_score_from` method | Big thumbs up!
Appropriately handles edge cases for tie-breaking logic | Good work on this. Clear flow, correct output!
All tests for `highest_score_from` pass | Yupp!
**Overall** |
